File: Codigo/Lectores-Escritores/readers_writers.c
#include <pthread.h>
#include <semaphore.h>
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>

sem_t sem_escritura;     // Controla el acceso a los libros
sem_t sem_turno;         // Garantiza el orden FIFO (justicia)
pthread_mutex_t mutex;   // Protege el conteo de lectores activos
int lectores_activos = 0;
int libros = 1;          // Recurso compartido: cantidad o estado de los libros

// ====== Variables para métricas ======
double *tiempos_espera;
int total_hilos;
int operaciones_completadas = 0;
pthread_mutex_t mutex_metricas;

// Función auxiliar para obtener tiempo en segundos
double get_time() {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec + ts.tv_nsec / 1e9;
}

void *lector(void *arg)
{
    int id = *((int *)arg);
    sleep(rand() % 3); // Simula llegada aleatoria

    double inicio_espera = get_time();

    sem_wait(&sem_turno); // Espera su turno (FIFO)
    pthread_mutex_lock(&mutex);
    lectores_activos++;
    if (lectores_activos == 1)
        sem_wait(&sem_escritura); // El primer lector bloquea a los escritores
    pthread_mutex_unlock(&mutex);
    sem_post(&sem_turno); // Libera el turno

    double fin_espera = get_time();
    tiempos_espera[id - 1] = fin_espera - inicio_espera;

    printf("📖 Lector %d leyó los libros = %d\n", id, libros);
    sleep(1); // Tiempo de lectura

    pthread_mutex_lock(&mutex);
    lectores_activos--;
    if (lectores_activos == 0)
        sem_post(&sem_escritura); // El último lector libera a los escritores
    pthread_mutex_unlock(&mutex);

    pthread_mutex_lock(&mutex_metricas);
    operaciones_completadas++;
    pthread_mutex_unlock(&mutex_metricas);

    return NULL;
}

void *escritor(void *arg)
{
    int id = *((int *)arg);
    sleep(rand() % 3); // Simula llegada aleatoria

    double inicio_espera = get_time();

    sem_wait(&sem_turno);     // Espera su turno (FIFO)
    sem_wait(&sem_escritura); // Espera permiso para escribir
    sem_post(&sem_turno);     // Libera el turno

    double fin_espera = get_time();
    tiempos_espera[id - 1] = fin_espera - inicio_espera;

    libros += 3; // Modifica el recurso
    printf("✍️  Escritor %d actualizó los libros a %d\n", id, libros);
    sleep(1); // Tiempo de escritura

    sem_post(&sem_escritura); // Libera el recurso

    pthread_mutex_lock(&mutex_metricas);
    operaciones_completadas++;
    pthread_mutex_unlock(&mutex_metricas);

    return NULL;
}

int main()
{
    int n_lectores, n_escritores;
    srand(time(NULL));

    printf("=== Simulación Lectores–Escritores (FIFO justa, con métricas) ===\n");
    printf("Ingrese número de lectores: ");
    scanf("%d", &n_lectores);
    printf("Ingrese número de escritores: ");
    scanf("%d", &n_escritores);

    total_hilos = n_lectores + n_escritores;
    tiempos_espera = malloc(sizeof(double) * total_hilos);

    pthread_t lectores[n_lectores], escritores[n_escritores];
    int id_lectores[n_lectores], id_escritores[n_escritores];

    pthread_mutex_init(&mutex, NULL);
    pthread_mutex_init(&mutex_metricas, NULL);
    sem_init(&sem_escritura, 0, 1);
    sem_init(&sem_turno, 0, 1);

    double inicio_total = get_time();

    // Crear hilos lectores
    for (int i = 0; i < n_lectores; i++) {
        id_lectores[i] = i + 1;
        pthread_create(&lectores[i], NULL, lector, &id_lectores[i]);
    }

    // Crear hilos escritores
    for (int i = 0; i < n_escritores; i++) {
        id_escritores[i] = n_lectores + i + 1;
        pthread_create(&escritores[i], NULL, escritor, &id_escritores[i]);
    }

    // Esperar finalización
    for (int i = 0; i < n_lectores; i++)
        pthread_join(lectores[i], NULL);
    for (int i = 0; i < n_escritores; i++)
        pthread_join(escritores[i], NULL);

    double fin_total = get_time();
    double duracion_total = fin_total - inicio_total;

    // ==== Cálculo de métricas ====
    double suma = 0, suma_cuadrados = 0;
    for (int i = 0; i < total_hilos; i++) {
        suma += tiempos_espera[i];
        suma_cuadrados += tiempos_espera[i] * tiempos_espera[i];
    }

    double promedio_espera = suma / total_hilos;
    double varianza = (suma_cuadrados / total_hilos) - (promedio_espera * promedio_espera);
    double fairness = sqrt(varianza);
    double throughput = operaciones_completadas / duracion_total;
    double overhead = promedio_espera / total_hilos;

    printf("\n=== MÉTRICAS DEL ESCENARIO ===\n");
    printf("⏱ Duración total: %.3f s\n", duracion_total);
    printf("📊 Tiempo promedio de espera por recurso: %.3f s\n", promedio_espera);
    printf("⚙️  Throughput: %.3f operaciones/s\n", throughput);
    printf("⚖️  Fairness (desviación estándar): %.3f s\n", fairness);
    printf("🔁 Overhead de sincronización: %.6f s\n", overhead);

    printf("\nCantidad final de libros: %d\n", libros);
    printf("=== Fin de la simulación ===\n");

    // Liberar recursos
    pthread_mutex_destroy(&mutex);
    pthread_mutex_destroy(&mutex_metricas);
    sem_destroy(&sem_escritura);
    sem_destroy(&sem_turno);
    free(tiempos_espera);

    return 0;
}

File: Codigo/Lectores-Escritores/readers-writers.py
#!/usr/bin/env python3
"""
Simulación Lectores–Escritores (FIFO justa) CON métricas.
Lectores y Escritores tienen IDs independientes (1..N cada uno).
Ejecutar: python3 readers-writers.py
"""

import threading
import time
import random
import math
import sys

# -----------------------
# Semáforos y locks
# -----------------------
sem_escritura = threading.Semaphore(1)  # controla acceso exclusivo a 'libros'
sem_turno = threading.Semaphore(1)      # garantiza un turno (cola simple)
mutex = threading.Lock()                # protege 'lectores_activos'
metricas_lock = threading.Lock()

# Estado del recurso
lectores_activos = 0
libros = 1

# -----------------------
# Variables de métricas (separadas por tipo)
# -----------------------
tiempos_espera_lectores = []   # index 0..n_lectores-1
tiempos_espera_escritores = [] # index 0..n_escritores-1
operaciones_completadas = 0

# helper para tiempo monotónico en segundos
def now():
    return time.monotonic()

# -----------------------
# Funciones de hilo
# -----------------------
def lector(arg_id):
    global lectores_activos, libros, operaciones_completadas

    id = arg_id
    # llegada aleatoria (simula llegada)
    time.sleep(random.randint(0, 2))

    inicio_espera = now()

    # Espera su turno FIFO
    sem_turno.acquire()
    with mutex:
        lectores_activos += 1
        if lectores_activos == 1:
            # primer lector bloquea a escritores
            sem_escritura.acquire()
    # libera turno
    sem_turno.release()

    fin_espera = now()
    # registrar en la lista de lectores (id relativo 1..n_lectores)
    tiempos_espera_lectores[id - 1] = fin_espera - inicio_espera

    # sección crítica (lectura)
    print(f"📖 Lector {id} leyó los libros = {libros}")
    time.sleep(1)  # tiempo de lectura

    # salida
    with mutex:
        lectores_activos -= 1
        if lectores_activos == 0:
            sem_escritura.release()

    # contabilizar operación terminada
    with metricas_lock:
        operaciones_completadas += 1


def escritor(arg_id):
    global libros, operaciones_completadas

    id = arg_id
    # llegada aleatoria
    time.sleep(random.randint(0, 2))

    inicio_espera = now()

    # cola FIFO
    sem_turno.acquire()
    sem_escritura.acquire()
    # libera turno (ahora el escritor tiene la exclusividad)
    sem_turno.release()

    fin_espera = now()
    # registrar en la lista de escritores (id relativo 1..n_escritores)
    tiempos_espera_escritores[id - 1] = fin_espera - inicio_espera

    # sección crítica (escritura)
    libros += 3
    print(f"✍️  Escritor {id} actualizó los libros a {libros}")
    time.sleep(1)  # tiempo de escritura

    # libera recurso
    sem_escritura.release()

    # contabilizar operación terminada
    with metricas_lock:
        operaciones_completadas += 1


# -----------------------
# Función principal
# -----------------------
def main():
    global tiempos_espera_lectores, tiempos_espera_escritores, operaciones_completadas, libros, lectores_activos

    # pedir número de lectores/escritores (igual que en C)
    try:
        n_lectores = int(input("Ingrese número de lectores: ").strip())
        n_escritores = int(input("Ingrese número de escritores: ").strip())
    except Exception:
        print("Entrada inválida. Usa números enteros.")
        sys.exit(1)

    # inicializar listas de métricas por tipo
    tiempos_espera_lectores = [0.0] * n_lectores
    tiempos_espera_escritores = [0.0] * n_escritores

    lectores_threads = []
    escritores_threads = []

    # reiniciar estado global como en C
    lectores_activos = 0
    libros = 1
    operaciones_completadas = 0

    # seed
    random.seed(int(time.time()))

    t_inicio_total = now()

    # crear lectores (IDs 1..n_lectores)
    for i in range(n_lectores):
        tid = i + 1
        th = threading.Thread(target=lector, args=(tid,))
        lectores_threads.append(th)
        th.start()

    # crear escritores (IDs 1..n_escritores)
    for i in range(n_escritores):
        tid = i + 1
        th = threading.Thread(target=escritor, args=(tid,))
        escritores_threads.append(th)
        th.start()

    # join
    for th in lectores_threads:
        th.join()
    for th in escritores_threads:
        th.join()

    t_fin_total = now()
    duracion_total = t_fin_total - t_inicio_total

    # ---- combinar las métricas para cálculo global ----
    tiempos_combinados = tiempos_espera_lectores + tiempos_espera_escritores
    total_hilos = len(tiempos_combinados)

    suma = sum(tiempos_combinados)
    suma_cuadrados = sum(x * x for x in tiempos_combinados)
    promedio_espera = suma / total_hilos if total_hilos > 0 else 0.0
    varianza = (suma_cuadrados / total_hilos - promedio_espera * promedio_espera) if total_hilos > 0 else 0.0
    fairness = math.sqrt(varianza) if varianza > 0 else 0.0
    throughput = operaciones_completadas / duracion_total if duracion_total > 0 else 0.0
    overhead = promedio_espera / total_hilos if total_hilos > 0 else 0.0

    print("\n=== MÉTRICAS DEL ESCENARIO ===")
    print(f"⏱ Duración total: {duracion_total:.3f} s")
    print(f"📊 Tiempo promedio de espera por recurso: {promedio_espera:.3f} s")
    print(f"⚙️  Throughput: {throughput:.3f} operaciones/s")
    print(f"⚖️  Fairness (desviación estándar): {fairness:.3f} s")
    print(f"🔁 Overhead de sincronización: {overhead:.6f} s")
    print(f"\nCantidad final de libros: {libros}")
    print("=== Fin de la simulación ===")

if __name__ == "__main__":
    main()
